File: Frog jump with k distance/Memoization.java
import java.util.Arrays;

public class Memoization {

    public static int FropJump(int size,int[] heights,int k,int[] dp) {
        if(dp[size] != -1) return dp[size];
        if(size == 0) return 0;
        int minimumJump = Integer.MAX_VALUE;

        for(int j = 1;j <= k ;j ++) {
            if(size - j >= 0) {
                int jump = FropJump(size-j, heights, k, dp) + Math.abs(heights[size] - heights[size-j]);
                minimumJump = Math.min(jump, minimumJump);
            }

            else {
                break;
            }
        }
        return dp[size] = minimumJump;
    }
    public static void main(String[] args) {
        int heights[] = {10, 5, 20, 0, 15};
        int size = heights.length;
        int dp[] = new int[size];
        int k = 2;
        Arrays.fill(dp, -1);

        System.out.println(FropJump(size-1,heights,k,dp));
    }
}

File: Frog jump with k distance/Memoization.py
def minimumJump(arr,size,k,dp):
    if dp[size] != -1: return dp[size]
    if size == 0: return 0
    minimum = float('inf')

    for j in range(1,k+1):
        if size - j >= 0:
            minimum = min(minimum,abs(arr[size] - arr[size-j]) + minimumJump(arr,size-j,k,dp))
        else: break 
    
    dp[size] = minimum
    return dp[size]

def FrogJump():
    arr = [10, 5, 20, 0, 15]
    size = len(arr)-1
    k = 2
    dp = [-1 for _ in range(size+1)]

    return minimumJump(arr,size,k,dp)

if __name__ == "__main__":
    print(FrogJump())